FEAT: implement jax-based runtime graph fusion optimizer

RandomY-2 opened the pull request:
<!--
Thank you for your contribution!
-->

## What do these changes do?

Add jax as an engine for runtime optimization. Currently, only a few basic arithmetic operations are supported, but jax should allow more types of nodes to be fused comparing to numexpr. I was able to pass all tests with engines set to ["jax", "cupy"] for https://github.com/xprobe-inc/xorbits/blob/47ceb99245b7c4d4c4372ee11e418e20942f090b/python/xorbits/_mars/optimization/physical/core.py#L89 and https://github.com/xprobe-inc/xorbits/blob/47ceb99245b7c4d4c4372ee11e418e20942f090b/python/xorbits/_mars/services/task/config.py#L27

Performance-wise, improved execution by 12.5%

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->
Fixes #xxxx

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

aresnow1 commented:
Awesome! We've been wanting to try doing this, great job!

aresnow1 commented:
You can add a workflow in ./github/workflows/python.yaml to test jax in CI.

ChengjieLi28 commented on python/xorbits/deploy/docker/Dockerfile.base:
```diff
@@ -23,13 +23,15 @@ ARG PYTHON_VERSION=3.9
     /opt/conda/bin/conda install --freeze-installed -c conda-forge \
     python=3.11 numpy\>=1.14.0 pandas\>=1.5.0 cython\>=0.29.33 scipy\>=1.9.2 \
-    mkl numexpr psutil scikit-learn sqlalchemy tornado lz4 ; fi
+    mkl numexpr jax jaxlib psutil scikit-learn sqlalchemy tornado lz4 ; fi
```
Installing ``jax`` and ``jaxlib`` on ``final`` docker layer would be enough. Is there any special reason to install it again under python 3.11 ?

RandomY-2 replied:
Hi, I just followed the installation steps for numexpr, which appears in both places. I am not sure if there is a special purpose to it, but I suppose I will remove this part then? Thanks.

ChengjieLi28 replied:
> Hi, I just followed the installation steps for numexpr, which appears in both places. I am not sure if there is a special purpose to it. So maybe I will remove this part then? Thanks.

Here's the thing, python 3.11 is a relatively new version of python, and the dependencies we need like pandas and numpy have unique requirements for this python version, so the 3.11 layer installs these dependencies separately while specifying the ``--freeze-installed`` option. ``jax`` and ``jaxlib`` should just install normally for each python version. And ``jax`` and ``jaxlib`` should be installed via the ``conda-forge`` channel according to ``conda search`` result.

Therefore, add ``jax`` and ``jaxlib`` to ``conda-forge`` installation on ``final`` layer (docker file line 39-42) would be good.

RandomY-2 replied:
I see, thanks, i have removed the install

RandomY-2 commented:
> You can add a workflow in ./github/workflows/python.yaml to test jax in CI.

Hi, how do I do that? 👀

aresnow1 commented:
> > You can add a workflow in ./github/workflows/python.yaml to test jax in CI.
> 
> Hi, how do I do that? 👀

We use GitHub Actions to run CI. The workflow is defined in `.github/workflows/python.yaml`. Each workflow includes steps such as setting up the environment, installing dependencies, building, testing, and uploading coverage. In this yaml file, we also differentiate between different tests using the `module` variable. So, you can add a job with the module as "jax" in the include section(https://github.com/xprobe-inc/xorbits/blob/main/.github/workflows/python.yaml#L84). Then, in the following steps, if the module is "jax," you can install the corresponding dependencies and run the jax tests. Basically, you just need to add some configurations in this file. For more details, you can refer to the official documentation(https://docs.github.com/en/actions).

UranusSeven commented on python/xorbits/_mars/tensor/fuse/jax.py:
```diff
@@ -0,0 +1,77 @@
+            _func = _evaluate(node)
+            funcs.append(_func)
+
+        def _fusion(inputs):
```
```suggestion
        def _fusion(*inputs):
```

In `execute`, you are passing `*inputs` as the argument, which means there could be more than 1 positional arguments.

RandomY-2 replied:
Hi, yes indeed the code was not correct. I updated the code, but it still has some issues with the output dimension. I will look tomorrow

UranusSeven commented on python/xorbits/_mars/tensor/fuse/jax.py:
```diff
@@ -0,0 +1,77 @@
+            funcs.append(_func)
+
+        def _fusion(inputs):
+            output = funcs[0](*inputs)
```
Here's how to reproduce the problem:

```python
def test(setup):
    rs = np.random.RandomState(0)
    raw1 = rs.randint(10, size=(10, 10, 10))
    raw2 = rs.randint(10, size=(10, 10, 10))
    arr1 = tensor(raw1, chunk_size=5)
    arr2 = tensor(raw2, chunk_size=5)

    arr3 = arr1 + arr2 + 10
    arr4 = 10 + arr1 + arr2
    res3 = arr3.execute().fetch()
    res3_cmp = arr4.execute().fetch()
    np.testing.assert_array_equal(res3, res3_cmp)
```

UranusSeven commented on python/xorbits/_mars/tensor/fuse/jax.py:
```diff
@@ -0,0 +1,107 @@
+from .core import TensorFuseChunkMixin
+
+
+class TensorJAXFuseChunk(TensorFuse, TensorFuseChunkMixin):
```
I think we need to figure out a way to test this class. Do you have any idea?

RandomY-2 replied:
Yeah I agree. I was thinking about disabling the numexpr engine during the jax workflow, so only jax and cupy will be used for optimization and we can thus test jax, but i haven't found an easy way to do that 🤔

aresnow1 replied:
Placen jax in front of numexpr in `engines`, and jax will be chosen firstly if it is installed. Numexpr is also be tested when jax is not installed.

UranusSeven commented on python/xorbits/_mars/tensor/fuse/jax.py:
```diff
@@ -0,0 +1,107 @@
+    @classmethod
+    def execute(cls, ctx, op):
+        chunk = op.outputs[0]
+        inputs = as_same_device([ctx[c.key] for c in op.inputs], device=op.device)
```
Now the inputs should have already been on the same device before execution. This is guaranteed by the fetch mechanism.
```suggestion
        inputs = [ctx[c.key] for c in op.inputs]
```

UranusSeven commented on python/xorbits/_mars/tensor/fuse/jax.py:
```diff
@@ -0,0 +1,107 @@
+
+
+# recursively compute the inputs
+def _evaluate(chunk, input_dict):
```
I benchmarked this function with the following test and found it performed 12.5% better than numexpr!!!

```python
def test_base_execution(setup):
    rs = np.random.RandomState(0)
    raw1 = rs.randint(10, size=(1000, 1000, 1000))
    raw2 = rs.randint(10, size=(1000, 1000, 1000))
    arr1 = tensor(raw1, chunk_size=50)
    arr2 = tensor(raw2, chunk_size=50)
    arr3 = arr1 + arr2 + 10
    res3 = arr3.execute().fetch()
```

I'm thinking if it is possible to generate a function like what we do with numexpr, so that we can get rid of the `if` statements and apply jit.

What do you think?

RandomY-2 replied:
yeah i can try doing that

UranusSeven replied:
We've decided to optimize this function later.

aresnow1 commented on .github/workflows/python.yaml:
```diff
@@ -183,6 +184,10 @@ jobs:
         fi
+        if [[ "$MODULE" == "jax" ]]; then
+          # test jax
+          pip install jax>=0.4.0 jaxlib>=0.4.0
```
I found they are not installed, try to add quotes like this, `pip install "jax>=0.4.0" "jaxlib>=0.4.0"`.

RandomY-2 commented:
Hi, do you have any suggestions on how to test the jax file? using it as the engine to run existing tests seems not enough... 🤔

RandomY-2 commented on python/xorbits/_mars/optimization/physical/core.py:
```diff
@@ -76,6 +87,150 @@ def _fuse_nodes(
+        node: ChunkType,
+        graph_results: Set[ChunkType],
+        cached_can_fuse,
+    ):
```
I have updated the type hints for the return values and some variables. could you check if those are sufficient now, thanks

aresnow1 commented:
> Hi, do you have any suggestions on how to test the jax file? using it as the engine to run existing tests seems not enough... 🤔

You can run tests locally to see if it invokes those code, I'm not sure if these tests works for jax, as jax's supported ops are different with numexpr's, maybe the chunk nodes generated by tests are even not be merged into a TensorJAXFuseChunk, so that it is not invoked and covered. Add some new cases if needed.

RandomY-2 commented:
> > Hi, do you have any suggestions on how to test the jax file? using it as the engine to run existing tests seems not enough... 🤔
> 
> You can run tests locally to see if it invokes those code, I'm not sure if these tests works for jax, as jax's supported ops are different with numexpr's, maybe the chunk nodes generated by tests are even not be merged into a TensorJAXFuseChunk, so that it is not invoked and covered. Add some new cases if needed.

I updated the supported ops and jax should support all numexpr ops except tree add and tree multiply now. I also tried to print in the execute function in TensorJaxFuseChunk as:

```python
@classmethod
    def execute(cls, ctx, op):
        chunk = op.outputs[0]
        inputs = [ctx[c.key] for c in op.inputs]
        input_dict = {c.key: i for c, i in zip(op.inputs, inputs)}
        res = np.array(_evaluate(chunk, input_dict))
        print('jax fused') // print here
        res = _maybe_keepdims(chunk, res)
        if chunk.ndim == 0 and res.ndim == 1 and res.size == 0:
            res = res.dtype.type(0)
        ctx[chunk.key] = res
```

and the print statement shows up in all six test cases when I run the /fuse tests with ```pytest -rP``` locally, so the code should be executed. Could you double-check if my workflow is set up correctly?

aresnow1 commented:
Codecov didn't find coverage files, try this, `pytest --cov-config=setup.cfg --cov-report=xml --cov=xorbits xorbits/_mars/tensor/fuse/tests/test_runtime_fusion.py`

RandomY-2 commented:
I have updated the type hints for the physical code and updated the jax tensor code so it should support the same ops as numexpr now. that said, generating the jit expression is still a bit tricky. could we save this part for a future pr? i am also a bit busy right now, so if someone is interested in doing the jit optimization part they can feel free to implement it.

aresnow1 commented:
> I have updated the type hints for the physical code and updated the jax tensor code so it should support the same ops as numexpr now. that said, generating the jit expression is still a bit tricky. could we save this part for a future pr? i am also a bit busy right now, so if someone is interested in doing the jit optimization part they can feel free to implement it.

This PR looks good to me now, generating jit function can be done in the following PRs.

aresnow1 reviewed (approved):
LGTM

UranusSeven reviewed (approved):
LGTM.
